src/main/java/aztech/modern_industrialization/machines/multiblocks/shape/ShapeTemplate.java

```diff
@@ -21,65 +21,88 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-package aztech.modern_industrialization.machines.multiblocks;
+package aztech.modern_industrialization.machines.multiblocks.shape;
 
+import aztech.modern_industrialization.MI;
 import aztech.modern_industrialization.machines.models.MachineCasing;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Map;
-import java.util.Set;
+import aztech.modern_industrialization.machines.multiblocks.shape.member.MultiblockMember;
+import aztech.modern_industrialization.machines.multiblocks.shape.member.SimpleMultiblockMember;
+import java.util.*;
 import net.minecraft.core.BlockPos;
+import net.neoforged.bus.api.SubscribeEvent;
+import net.neoforged.fml.common.EventBusSubscriber;
+import net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * An immutable description of a multiblock shape.
  */
-public class ShapeTemplate {
-    public final Map<BlockPos, SimpleMember> simpleMembers = new HashMap<>();
-    public final Map<BlockPos, HatchFlags> hatchFlags = new HashMap<>();
```

Comment:
I wouldn't make the hatch flags inherit from `SimpleMultiblockMember`, that's messy. We could group both a member and hatch flags into a record.

Reply:
The reason for grouping the hatch flags into `HatchMultiblockMember` is so the hatches can be serialized as part of the structure via. user input in a GUI much easier. I know this introduces a few `instanceof` checks here and there which isn't ideal. Other than that I'm not sure I agree that it's messy... what's your reasoning?